doc/design/proposal-synchronous-hooks.md

```diff
@@ -0,0 +1,258 @@
+
+For background and motivation, see https://github.com/nodejs/node/issues/52219. Prototype implementation is in https://github.com/joyeecheung/node/tree/sync-hooks.
+
+TL;DR: the top priority of this proposal is to allow sun-setting CJS loader monkey patching as non-breakingly as possible (which is why it needs to be synchronous and in-thread because that's how the CJS loader and `require()` works). Then it's API consistency with the existing `module.register()` off-thread hooks.
+
+Existing users of CJS loader monkey patching to look into:
+
+- pirates (used by Babel and nyc)
+- require-in-the-middle (used by many tracing agents)
+- yarn pnp
+- tsx
+- ts-node
+
+## API design
+
+A high-level overview of the API.
+
+```js
+const { addHooks, removeHooks } = require('module');
+
+function resolve(specifier, context, nextResolve) {
+  const resolved = nextResolve(specifier, context);
+  return { url: resolved.url.replaceAll(/foo/g, 'bar'); }
+}
+
+function load(url, context, nextLoad) {
+  const loaded = nextLoad(specifier, context);
+  return { source: loaded.source.toString().replaceAll(/foo/g, 'bar'); }
+}
+
+// id is a symbol
+const id = addHooks({ resolve, load, ... });
+removeHooks(id);
+```
+
+1. The names `addHooks` and `removeHooks` take inspiration from pirates. Can be changed to other better names.
+2. An alternative design to remove the hooks could be `addHooks(...).unhook()`, in this case `addHooks()` returns an object that could have other methods.
+   1. This may allow third-party hooks to query itself and avoid double-registering. Though this functionality is probably out of scope of the MVP.
+3. Another alternative design would be like what pirates offer: `const revert = addHooks(..); revert();`.
+4. It seems useful to allow the results returned by the hooks to be partial i.e. hooks don't have to clone the result returned by the next (default) hook to override it, instead they only need to return an object with properties that they wish to override. This can save the overhead of excessive clones.
+5. It seems `shortCircuit` is not really necessary if hooks can just choose to not call the next hook?
+
+## Hooks
+
+### `resolve`: from specifier to url
+
+```js
+/**
+ * @typedef {{
+ *   parentURL?: string,
+ *   conditions?: string[],
+ *   importAttributes?: object[]
+ * }} ModuleResolveContext
+ */
+/**
+ * @typedef {{
+ *   url?: string,
+ *   format?: string
+ * }} ModuleResolveResult
+ */
+/**
+ * @param {string} specifier
+ * @param {ModuleResolveContext} context
+ * @param {(specifier: string, context: ModuleResolveContext) => ModuleResolveResult} nextResolve
+ * @returns {ModuleResolveResult}
+ */
+function resolve(specifier, context, nextResolve) {
+  if (shouldOverride(specifier, context.parentURL)) {
+    const url = customResolve(specifier, context.parentURL);
+    return { url };
+  }
+
+  const resolved = nextResolve(specifier, context);
+  if (resolved.url.endsWith('.zip')) {
+    return { format: 'zip' };
```

Comment:
That only matters if all properties are required rather than what Joyee was suggesting of not needing to copy unnecessarily and only including what changes.